File: breakout_1/ATP_coveredgr/js-jest/test/gilded_rose.test.js
const {Shop, Item} = require("../src/gilded_rose");

describe("Gilded Rose", function () {
    it("Legendary Item quality does not decrease", () => {
        // Arrange
        const sut = new Shop([new Item("Sulfuras, Hand of Ragnaros", 20, 80)]);

        // Act
        const items = sut.updateQuality();

        // Assert
        expect(items[0].quality).toBe(80);
    });

    it("Legendary Item never has to be sold", () => {
        const sut = new Shop([new Item("Sulfuras, Hand of Ragnaros", 1, 80)]);

        const items = sut.updateQuality();

        expect(items[0].sellIn).toBe(1);
    });

    it("Legendary Item quality does not change after sellin", () => {
        const sut = new Shop([new Item("Sulfuras, Hand of Ragnaros", -1, 80)]);

        const items = sut.updateQuality();

        expect(items[0].quality).toBe(80);
    });

    it.each([
        ["generic item"],
        ["Aged Brie"],
        ["Backstage passes to a TAFKAL80ETC concert"],
    ])('%p item SellIn decreases each update', (itemName) => {
        const sut = new Shop([new Item(itemName, 8, 10)]);

        const items = sut.updateQuality();

        expect(items[0].sellIn).toBe(7);
    });

    it.each([
        ["generic item"],
        ["Aged Brie"],
        ["Backstage passes to a TAFKAL80ETC concert"],
    ])('%p item  SellIn will be negative after sellIn date reached', (itemName) => {
        const sut = new Shop([new Item(itemName, 0, 25)]);

        const items = sut.updateQuality();

        expect(items[0].sellIn).toBe(-1);
    });

    it("Generic item quality decreases by 1 before sellIn date reached", () => {
        const sut = new Shop([new Item("generic item", 5, 10)]);

        const items = sut.updateQuality();

        expect(items[0].quality).toBe(9);
    });

    it("Generic item quality decreases twice as fast after sellIn date reached", () => {
        const sut = new Shop([new Item("generic item", 0, 10)]);

        const items = sut.updateQuality();

        expect(items[0].quality).toBe(8);
    });

    it("Generic item quality never decreases below 0", () => {
        const sut = new Shop([new Item("generic item", 5, 0)]);

        const items = sut.updateQuality();

        expect(items[0].quality).toBe(0);
    });

    it("Aged Brie quality improves with age", () => {
        const sut = new Shop([new Item("Aged Brie", 5, 30)]);

        const items = sut.updateQuality();

        expect(items[0].quality).toBe(31);
    });

    it("Aged Brie quality improves twice as fast after sellIn date reached", () => {
        const sut = new Shop([new Item("Aged Brie", 0, 30)]);

        const items = sut.updateQuality();

        expect(items[0].quality).toBe(32);
    });

    it("Aged Brie quality capped at 50 even when really old", () => {
        const sut = new Shop([new Item("Aged Brie", -10, 30)]);

        const items = sut.updateQuality();

        expect(items[0].quality).toBe(32);
    });

    it.each([
        ["Aged Brie"],
        ["Backstage passes to a TAFKAL80ETC concert"],
    ])('%p item that improves with age, quality is capped at 50 before sellIn date reached', (itemName) => {
        const sut = new Shop([new Item(itemName, 10, 50)]);

        const items = sut.updateQuality();

        expect(items[0].quality).toBe(50);
    });

    it("Backstage pass quality increases when concert is far in the future", () => {
        const sut = new Shop([new Item("Backstage passes to a TAFKAL80ETC concert", 30, 23)]);

        const items = sut.updateQuality();

        expect(items[0].quality).toBe(24);
    });

    it("Backstage pass quality increases more when concert is 10 days away", () => {
        const sut = new Shop([new Item("Backstage passes to a TAFKAL80ETC concert", 10, 23)]);

        const items = sut.updateQuality();

        expect(items[0].quality).toBe(25);
    });

    it("Backstage pass quality increases even more when concert is 5 days away", () => {
        const sut = new Shop([new Item("Backstage passes to a TAFKAL80ETC concert", 5, 23)]);

        const items = sut.updateQuality();

        expect(items[0].quality).toBe(26);
    });

    it("Backstage pass quality drops to zero when concert has passed", () => {
        const sut = new Shop([new Item("Backstage passes to a TAFKAL80ETC concert", 0, 23)]);

        const items = sut.updateQuality();

        expect(items[0].quality).toBe(0);
    });

    it("Backstage pass quality is capped at 50 when concert is about to happen", () => {
        const sut = new Shop([new Item("Backstage passes to a TAFKAL80ETC concert", 1, 49)]);

        const items = sut.updateQuality();

        expect(items[0].quality).toBe(50);
    });

    it("Shop can contain multiple items and each is updated", () => {
        const sut = new Shop([new Item("Sulfuras, Hand of Ragnaros", 0, 80),
        new Item("generic item", 10, 5)]);

        const items = sut.updateQuality();

        // jest doesn't have a mechanism to ensure all expectations are verified
        expect(items[0].sellIn).toBe(0);
        expect(items[1].sellIn).toBe(9);
    });

});

File: breakout_1/ATP_coveredgr/csharp/GildedRose/GildedRoseTest.cs
using NUnit.Framework;
using System.Collections.Generic;

namespace GildedRose
{
    [TestFixture]
    public class GildedRoseTest
    {
        private const string BackstagePass = "Backstage passes to a TAFKAL80ETC concert";

        private List<Item> createItemList(string itemName, int sellIn, int quality) {
            return new List<Item> { new Item { Name = itemName, SellIn = sellIn, Quality = quality } };
        }        

        [Test]
        public void LegendaryItem_QualityDoesNotDecrease() {
            // Arrange
            GildedRose sut = new GildedRose(createItemList("Sulfuras, Hand of Ragnaros", 20, 80));
    
            // Act
            sut.UpdateQuality();
    
            // Assert
            Assert.AreEqual(80, sut.Items[0].Quality, "Quality is not decreased for this legendary item");
        }
    
        [Test]
        public void LegendaryItem_NeverHasToBeSold() {
            GildedRose sut = new GildedRose(createItemList("Sulfuras, Hand of Ragnaros", 1, 80));
            sut.UpdateQuality();
            Assert.AreEqual(1, sut.Items[0].SellIn, "Sellin is not decreased for this legendary item");
        }

        [TestCase("generic item")]
        [TestCase("Aged Brie")]
        [TestCase(BackstagePass)]
        public void NonLegendaryItem_SellInDate_Decreases(string name) {
            GildedRose sut = new GildedRose(createItemList(name, 8, 10));
            sut.UpdateQuality();
            Assert.AreEqual(7, sut.Items[0].SellIn, "Item sellin date should decrease by 1 each day");
        }
    
        [TestCase("generic item")]
        [TestCase("Aged Brie")]
        [TestCase(BackstagePass)]
        public void NonLegendaryItem_SellInDate_CanBeNegative(string name) {
            GildedRose sut = new GildedRose(createItemList(name, 0, 25));
            sut.UpdateQuality();
            Assert.AreEqual(-1, sut.Items[0].SellIn, "Sellin date will go negative once sellin date is reached");
        }
    
        [Test]
        public void GenericItem_QualityDecreasesBeforeSellinDate() {
            GildedRose sut = new GildedRose(createItemList("generic item", 5, 10));
            sut.UpdateQuality();
            Assert.AreEqual(9, sut.Items[0].Quality, "Item quality should only decrease by 1 each day");
        }
    
        [Test]
        public void GenericItem_QualityDecreasesTwiceAsFastAfterSellinDate() {
            GildedRose sut = new GildedRose(createItemList("generic item", 0, 10));
            sut.UpdateQuality();
            Assert.AreEqual(8, sut.Items[0].Quality, "When sellin date is 0 then quality decreases twice as fast");
        }
    
        [Test]
        public void GenericItem_QualityNeverGoesNegative() {
            GildedRose sut = new GildedRose(createItemList("generic item", 0, 0));
            sut.UpdateQuality();
            Assert.AreEqual(0, sut.Items[0].Quality, "Quality will not go negative once it is zero");
        }
    
        [Test]
        public void AgedBrie_QualityIncreases() {
            GildedRose sut = new GildedRose(createItemList("Aged Brie", 5, 30));
            sut.UpdateQuality();
            Assert.AreEqual(31, sut.Items[0].Quality, "Aged Brie increases quality with age");
        }
    
        [TestCase("Aged Brie")]
        [TestCase(BackstagePass)]
        public void NonLegendaryItem_ThatImprovesWithAge_QualityIsCappedAt50(string itemName) {
            GildedRose sut = new GildedRose(createItemList(itemName, 5, 50));
            sut.UpdateQuality();
            Assert.AreEqual(50, sut.Items[0].Quality, "Quality has an upper limit that is not exceeded");
        }
    
        [Test]
        public void AgedBrie_QualityIncreases_EvenAfterSellInDate() {
            GildedRose sut = new GildedRose(createItemList("Aged Brie", -1, 20));
            sut.UpdateQuality();
            Assert.AreEqual(22, sut.Items[0].Quality, "Aged Brie improves twice as fast after sellin date (BUG?)");
        }
    
        [Test]
        public void AgedBrie_QualityIsCappedAt50_EvenWhenReallyOld() {
            GildedRose sut = new GildedRose(createItemList("Aged Brie", -99, 50));
            sut.UpdateQuality();
            Assert.AreEqual(50, sut.Items[0].Quality, "Quality has an upper limit, even when cheese is old");
        }
    
        [Test]
        public void BackstagePass_QualityIncreasesEachDay() {
            GildedRose sut = new GildedRose(createItemList(BackstagePass, 30, 23));
            sut.UpdateQuality();
            Assert.AreEqual(24, sut.Items[0].Quality, "Backstage Pass increases quality with age");
        }
    
        [Test]
        public void BackstagePass_QualityIncreasesMoreAsConcertNears() {
            GildedRose sut = new GildedRose(createItemList(BackstagePass, 10, 40));
            sut.UpdateQuality();
            Assert.AreEqual(42, sut.Items[0].Quality, "Backstage Pass quality increases more when concert is near");
        }
    
        [Test]
        public void BackstagePass_QualityIncreasesMuchMoreWhenConcertIsClose() {
            GildedRose sut = new GildedRose(createItemList(BackstagePass, 5, 40));
            sut.UpdateQuality();
            Assert.AreEqual(43, sut.Items[0].Quality, "Backstage Pass quality increases even more when concert is almost here");
        }
    
        [Test]
        public void BackStagePass_QualityDropsToZeroWhenConcertPasses() {
            GildedRose sut = new GildedRose(createItemList(BackstagePass, 0, 50));
            sut.UpdateQuality();
            Assert.AreEqual(0, sut.Items[0].Quality, "Backstage Pass is worthless when concert has passed");
        }

        [Test]
        public void ShopContainsMultipleItems() {
            List<Item> items = new List<Item> {
                new Item {Name = "Sulfuras, Hand of Ragnaros", SellIn = 0, Quality = 80},
                new Item {Name = "generic item", SellIn = 10, Quality = 5}
            };
            GildedRose sut = new GildedRose(items);
            
            sut.UpdateQuality();
            
            Assert.Multiple(() => {
                Assert.AreEqual(0, items[0].SellIn, "Legendary item SellIn date is not changed");
                Assert.AreEqual(9, items[1].SellIn, "generic item SellIn is decreased");
            });
        }
        
        //NEW BEHAVIOR
        // conjured items
    
    }
}
